gate-proxy/src/main/kotlin/com/netflix/spinnaker/gate/controllers/ProxyController.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.controllers
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.squareup.okhttp.Request
+import org.springframework.web.bind.annotation.*
+import org.springframework.web.servlet.HandlerMapping
+import javax.servlet.http.HttpServletRequest
+
+import com.netflix.spinnaker.config.ProxyConfigurationProperties
+import com.netflix.spinnaker.kork.web.exceptions.NotFoundException
+
+@RestController
+@RequestMapping(value = ["/proxies"])
+class ProxyController(val objectMapper: ObjectMapper,
+                      val proxyConfigurationProperties: ProxyConfigurationProperties) {
+  @GetMapping(value = ["/{proxy}/**"])
+  fun get(@PathVariable(value = "proxy") proxy: String,
+          @RequestParam requestParams: Map<String, String>,
+          httpServletRequest: HttpServletRequest): GenericResponse {
+
+    val proxyConfig = proxyConfigurationProperties
+      .proxies
+      .find { it.id.equals(proxy, true) } ?: throw NotFoundException("No proxy config found with id '$proxy'")
+
+    val proxyPath = httpServletRequest
+      .getAttribute(HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE)
+      .toString()
+      .substringAfter("/proxies/$proxy")
+
+    val proxiedUrlBuilder = Request.Builder().url(proxyConfig.uri + proxyPath).build().httpUrl().newBuilder()
+    for ((key, value) in requestParams) {
+      proxiedUrlBuilder.addQueryParameter(key, value)
+    }
+    val proxiedUrl = proxiedUrlBuilder.build()
+
+    val response = proxyConfig.okHttpClient.newCall(
+      Request.Builder().url(proxiedUrl).build()
+    ).execute()
+
+    val responseBody = response.body().string()
+    val responseObj = if (responseBody.startsWith("{")) {
+      objectMapper.readValue(responseBody, Map::class.java)
+    } else if (responseBody.startsWith("[")){
+      objectMapper.readValue(responseBody, Collection::class.java)
+    } else {
+      response.body().string()
+    }
+
+    return GenericResponse.ok(mapOf(
```

Comment:
not particularly tied to this structure ... I do sorta want to have a response wrapper tho

Reply:
I think I'd prefer to consume this without the response wrapper

Reply:
Fair point.

After a bit of thinking ... I'll remove the wrapper for now to keep it consistent with our other APIs. 

I hold out some hope that a future rev of our APIs will see us align on a consistent wrapper but we can cross that bridge late.

Reply:
What's the advantage of a wrapper around proxied calls to external services?

Reply:
I'd also vote to not wrap request/response objects for proxied services. It'd be sorta awkward in the case where the proxied service also has its own wrappers.

Reply:
One thing I'm not seeing in this code, however, is correct handling of proxied status codes. In addition to correctly handling proxied 5xx statuses as `502`, we should probably include `X-Proxy-Status-Code` which captures the underlying status code.

Reply:
Yar. Agree it doesn't make sense for straight up proxied services. 

@anotherchrisberry we can converse about why wrappers make (or might make) sense for other places where we're currently returning collection.

@robzienert Ya, I've not done anything yet with status code propagation ... but will.

Reply:
Totally agree with wrappers around collections from our internal services - just not crazy about changing the shape of external responses unless there's a good reason to do so.